### DISCUSSION TITLE
Crowd division

### DISCUSSION THREAD

**Discussion by FrameMuse** [isaac-mason/recast-navigation-js#433] [General]
## Description
I would like to see a crowd to divide when someone goes through it.

## Visualization
Imagine you have multiple agents forming two groups
![image](https://github.com/isaac-mason/recast-navigation-js/assets/53980482/36148164-02d5-43e6-8127-97e3ca90a962)

And then someone decided to go to the point that lies through these groups.
![image](https://github.com/isaac-mason/recast-navigation-js/assets/53980482/c246cbdc-5056-4fa7-989d-732ed4e29a92)

The behaviour I would like to see is these groups of agents would step off a bit to let an agent to go through
![image](https://github.com/isaac-mason/recast-navigation-js/assets/53980482/c97ca593-1c18-4fb3-a2d7-7467dd0ad8ca)

And then close up again.
![image](https://github.com/isaac-mason/recast-navigation-js/assets/53980482/5440b408-49c7-4b0c-883d-58a7a8126fe3)

## Research

I already investigate and played with `collisionQueryRange`, `obstacleAvoidanceType`, `separationWeight`, `updateFlags`, but couldn't achieve the desired behaviour.

---

What would you suggest to do in such situation? I would also like to do this with multiple agents at the same time and having some of the agents to be fixed (but probably this simply destroying the agent and creating an obstacle). Though I would love to hear your opinions, thanks ^-^

**Comment by isaac-mason** [comment]
Hi @FrameMuse,

I'd need to look at how detour handles agent steering a bit more closely before I can answer this completely.

Agent velocity is influenced by a target velocity and an obstacle avoidance velocity.

You can debug these with:
- https://docs.recast-navigation-js.isaacmason.com/classes/index.CrowdAgent.html#desiredVelocity
- https://docs.recast-navigation-js.isaacmason.com/classes/index.CrowdAgent.html#desiredVelocityObstacleAdjusted
- https://docs.recast-navigation-js.isaacmason.com/classes/index.CrowdAgent.html#velocity

If you're not able to tune obstacle avoidance parameters to get the behaviour you want, you may need to write some of your own logic for this. For example, you could pick a target or apply velocity for agents avoiding another agent, considering the other agents current path and position. 

Regarding static agents, creating an obstacle with the size of the agent is a good approach. It should also possible to set agent parameters so that they won't move. e.g. set their max acceleration and speed to 0.

Please share how you get on with this 🙂 curious to hear what you come up with

**Comment by FrameMuse** [comment]
This is what my experiments results are:

- I can't determine the steering behaviour against desired units. I can only set how an agent behaves against **everything**.
- If I set max acceleration and speed to 0 when near targets, they just "fly" past their targets and endlessly moving away.
- `obstacleAvoidanceType` and `queryFilterType` didn't have any effect on steering, at least I didn't notice anything.

So I will try to apply custom behaviour by checking agents in a radius, if someone is on the `nextTargetInPath`, I will apply acceleration to them perpendicularly a moving agent. Which should have the effect. But I have problems with that as when I use `requestMoveVelocity`, the target is resetting.

---

My "dream" is to make it very similar to what Company Of Heroes 1 or Supreme Commander 2 have.

**Comment by isaac-mason** [comment]
Thanks for sharing!

I believe using requestMoveVelocity overrides the "move request" set by requestMoveTarget yep. You would need to keep track of the previously set target yourself.

Keen to hear how you get on creating the custom behaviour.
